plugin/core/sessions.py

```diff
@@ -1232,7 +1251,9 @@ def end_async(self) -> None:
         if self.exiting:
             return
         self.exiting = True
-        self._plugin = None
+        if self._plugin:
+            self._plugin.on_session_end_async()
+            self._plugin = None
```

Comment:
Wouldn't you be able to detect a session ending via the `__del__` of your `AbstractPlugin`?

Reply:
This is so that I can clear all references to the class so that the `__del__` is actually triggered.

That said, I have some bug in the current implementation that prevents it from being called, even with this API. Need to look into it.

As for this API vs. weak references, I kinda prefer being explicit because otherwise, it's easy to forget to add stuff that is not weak referenced and prevent destructor from being called, without realizing it.

But if you insist, I could get rid of it and change to weak references.

Reply:
Found the issue with my code.

So then the question is: weak references or explicit API "on_session_end"?

I wouldn't like having to use weak references in this case. Using it internally (like within the LSP package) is fine but for an API it's not very explicit or clear. I'd need to make everything that stores reference to a class its method a weak pointer and then null-check every time it's used. If I have this API I know explicitly when things are available and when not (after `on_session_end_async`) and don't need extra null-checks everywhere. Also, an API like that is a good reminder to do things the right way to avoid leaks.

This would allow making a `session` pointer that would have an explicitly defined life rather than needing to use `weaksession` and having to null check it.

Reply:
(In that case, I would also need to call it in case of a crash)

Reply:
> Also, an API like that is a good reminder to do things the right way to avoid leaks.

I don't understand the context here, can you explain?

> So then the question is: weak references or explicit API "on_session_end"?

I don't know what's wisdom... I guess we can go ahead with `on_session_end`. Although I'm now wondering whether cyclic references are the cause of https://github.com/sublimelsp/LSP/issues/1456

Reply:
> > Also, an API like that is a good reminder to do things the right way to avoid leaks.
> 
> I don't understand the context here, can you explain?

In my case, the plugin creates an instance of `FileWatcher` which is passed a function to the Plugin's method. This connection needs to be broken on session ending or the class will not be garbage-collected.

Or if I want to for instance create an instance of the ApiWrapper class (which takes plugin instance as an argument) and store it in `self.api`, for example, then I need to clear out that property on session ending or it will also block the plugin instance from being GC'ed.

All those can be solved by using weakref but I'm not a fan of solving it that way. Also, using weakref enforces its use down the call chain (for example in this case within the ApiWrapper class) which seems excessive. And what would I do if I would have to work with a third-party API that I can't change?

Reply:
Alright, I’ll trust your judgement on it